### PR TITLE
File Types Order & Type Contents Order

### LINKED ISSUE #2294: Rule Request: File Content Order
This is a rule request. I'm willing to implement this myself and send a PR, but before doing that I'd like to know if the rule is wanted or not. 

## Rationale

When working in a team on a project, it can speed up both writing and especially finding code when all team members order things in a file the same way. While there may be cases where this is an anti-pattern to enable domain-specific grouping of things within a type or file, in many projects it can be useful to declare that a type should always represent one domain and be placed within its own file and things within a class shouldn't be grouped by domain, instead one should create additional types if more grouping is needed. This way types can be kept small and represent only one domain – and the order of things within a type can be checked more easily to ensure things are kept clean. This can also be a huge help in open-source projects with many contributors to ensure consistent style.

The requested rules should allow specifying an order for different types of things that can be part of a file and warnings should be shown if the order is broken somewhere. Here's a list of "type of things" a file could include:

- Supporting Types (outside the main type body)
- Cases (Enum only)
- Type Aliases
- Subtypes
- Stored Type Properties
- Computed Type Properties
- Stored Instance Properties
- Computed Instance Properties
- IBOutlets
- Initializers
- Type Methods
- Life-Cycle Methods
- IBActions
- Other Methods
- Subscripts
- Extensions (outside the main type body)

## Example

For example, consider this view controller which includes most of the above listed things:

```swift
// Supporting Types
protocol TestViewControllerDelegate {
    func didPressTrackedButton()
}

class TestViewController: UIViewController {
    // Type Aliases
    typealias CompletionHandler = ((TestEnum) -> Void)

    // Subtypes
    class TestClass {
        // 10 lines
    }

    struct TestStruct {
        // 3 lines
    }

    enum TestEnum {
        // 5 lines
    }

    // Stored Type Properties
    static let cellIdentifier: String = "AmazingCell"

    // Stored Instance Properties
    var shouldLayoutView1: Bool!
    weak var delegate: TestViewControllerDelegate?
    private var hasLayoutedView1: Bool = false
    private var hasLayoutedView2: Bool = false

    // Computed Instance Properties
    private var hasAnyLayoutedView: Bool {
         return hasLayoutedView1 || hasLayoutedView2
    }

    // IBOutlets
    @IBOutlet private var view1: UIView!
    @IBOutlet private var view2: UIView!

    // Initializers
    override init(nibName nibNameOrNil: String?, bundle nibBundleOrNil: Bundle?) {
        super.init(nibName: nibNameOrNil, bundle: nibBundleOrNil)
    }

    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }

    // Type Methods
    static func makeViewController() -> TestViewController {
        // some code
    }

    // Life-Cycle Methods
    override func viewDidLoad() {
        super.viewDidLoad()

        view1.setNeedsLayout()
        view1.layoutIfNeeded()
        hasLayoutedView1 = true
    }

    override func viewDidLayoutSubviews() {
        super.viewDidLayoutSubviews()

        view2.setNeedsLayout()
        view2.layoutIfNeeded()
        hasLayoutedView2 = true
    }

    // IBActions
    @IBAction func goNextButtonPressed() {
        goToNextVc()
        delegate?.didPressTrackedButton()
    }

    @objc
    func goToRandomVcButtonPressed() {
        goToRandomVc()
    }

    // MARK: Other Methods
    func goToNextVc() { /* TODO */ }

    func goToInfoVc() { /* TODO */ }

    func goToRandomVc() {
        let viewCtrl = getRandomVc()
        present(viewCtrl, animated: true)
    }

    private func getRandomVc() -> UIViewController { return UIViewController() }

    // Subscripts
    subscript(_ someIndexThatIsNotEvenUsed: Int) -> String {
        get {
            return "This is just a test"
        }

        set {
            log.warning("Just a test", newValue)
        }
    }
}

// Extensions
extension TestViewController: UITableViewDataSource {
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return 1
    }

    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        return UITableViewCell()
    }
}
```

This type is sorted in a way which I suggest as the default order unless something else is configured.

## Customizability

I think these rules must be opt-in and highly configurable. For example a basic configuration could look something like this:

```swift
file_types_order:
  custom:
    - supporting_types
    - main_type
    - extensions

type_contents_order:
  custom:
    - cases
    - type_aliases
    - subtypes
    - stored_type_properties
    - computed_type_properties
    - stored_instance_properties
    - computed_instance_properties
    - iboutlets
    - initializers
    - type_methods
    - life_cycle_methods
    - ibactions
    - other_methods
    - subscripts
```

Since some people might not want to differentiate between specific things, for example some might opt to mix up the order of stored and computed properties, one could also specify this:

```swift
file_types_order:
  custom:
    - supporting_types
    - main_type
    - extensions

type_contents_order:
  custom:
    - cases
    - type_aliases
    - subtypes
    - [stored_type_properties, computed_type_properties]
    - [stored_instance_properties, computed_instance_properties]
    - iboutlets
    - initializers
    - type_methods
    - life_cycle_methods
    - ibactions
    - other_methods
    - subscripts
```

This would allow stored and computed properties to be listed in mixed order, but still ensure properties are sorted below subtypes and above IBOutlets.

---

What do you think, would such a rule be added to SwiftLint? Are you missing any kinds of things within a subtype? Do you have other suggestions for the configuration? Any other ideas for the rules name?

I'm looking forward to hearing from you! 🙂

### PR DESCRIPTION
**Pull request by Dschee**
This fixes #2294 by introducing two new rules:
- `file_types_order`: Specifies how the types within a file should be ordered.
- `type_contents_order`: Specifies the order of subtypes, properties, methods & more within a type.

### CODE REVIEW

**Comment by Dschee** [comment]
I just added configuration files to both rules, this means the only missing thing is fixing the locations in the tests. Since I have no idea why this happens, help is highly appreciated! @jpsim @marcelofabri @norio-nomura

**Comment by marcelofabri** [comment]
> As you can see the offset to the init method points to somewhere right in the middle of the override keyword, which is weird.

I think it actually points to `init`, because we remove the violation marker from the string before passing to the rule.

**Comment by Dschee** [comment]
@marcelofabri Okay, you were right about that example which I fixed now. But that was not what was irritating me, here's a few examples which now still are a problem and I can't understand why:

```
/* 👨‍👩‍👧‍👦👨‍👩‍👧‍👦👨‍👩‍👧‍👦 */
↓class TestViewController: UIViewController {}
                                          ^ warning: File Types Order Violation: A 'main_type' should not be placed amongst the file type(s) 'supporting_type'. (file_types_order)

// Supporting Types
protocol TestViewControllerDelegate {
    func didPressTrackedButton()
}
```

```
/* 👨‍👩‍👧‍👦👨‍👩‍👧‍👦👨‍👩‍👧‍👦 */
// Supporting Types
protocol TestViewControllerDelegate {
    func didPressTrackedButton()
}

// Extensions
↓extension TestViewController: UITableViewDataSource {
                                          ^ warning: File Types Order Violation: A 'extension' should not be placed amongst the file type(s) 'main_type'. (file_types_order)
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return 1
    }

    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        return UITableViewCell()
    }
}

class TestViewController: UIViewController {}

// Extensions
extension TestViewController: UITableViewDataSource {
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return 1
    }

    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        return UITableViewCell()
    }
}
```

```
/* 👨‍👩‍👧‍👦👨‍👩‍👧‍👦👨‍👩‍👧‍👦 */
class TestViewController: UIViewController {
    // Subtypes
    ↓class TestClass {
        // 10 lines
    }
    ^ warning: Type Contents Order Violation: A 'subtype' should not be placed amongst the type content(s) 'type_alias,associated_type'. (type_contents_order)

    // Type Aliases
    typealias CompletionHandler = ((TestEnum) -> Void)
}
```

```
/* 👨‍👩‍👧‍👦👨‍👩‍👧‍👦👨‍👩‍👧‍👦 */
class TestViewController: UIViewController {
    // IBActions
    @IBAction ↓func goNextButtonPressed() {
        goToNextVc()
             ^ warning: Type Contents Order Violation: A 'ib_action' should not be placed amongst the type content(s) 'view_life_cycle_method'. (type_contents_order)
        delegate?.didPressTrackedButton()
    }

    // View Life-Cycle Methods
    override func viewDidLoad() {
        super.viewDidLoad()

        view1.setNeedsLayout()
        view1.layoutIfNeeded()
        hasLayoutedView1 = true
    }
}
```

It only happens when there's the `/* 👨‍👩‍👧‍👦👨‍👩‍👧‍👦👨‍👩‍👧‍👦 */` prepended at the beginning.

**Comment by Dschee** [comment]
Another problem I found was that when I uninstall SwiftLint entirely from my computer and run `make install` on my fork, then the new rules seem to work and find violations, but the `reason` statement isn't shown. As you can see in the examples above I have added specific reasons to explain what is placed amongst which things. This seems to work in tests, but when I compile the code into the CLI tool then in my project the violations simply state the `description` of the `RuleDescription`, not the specific `reason`. Is there something I need to do before this works?

**Comment by Dschee** [comment]
I just fixed all issues and rebased this to the lastest master. Ready to merge from my point of view!

**Comment by Dschee** [comment]
Okay, I just found the issue which made the CI fail when running this with Swift 4: It's due to `subscript` and `typealias` missing from SourceKit for versions below 4.1. I fixed this issue by not testing the cases which include the supporting type on Swift 4.

I hope the CI will be happy now and this can be merged. Anything else holding this off?

**Comment by Dschee** [comment]
Just rebased to solve merge conflicts.

**Comment by Dschee** [comment]
Just rebased this branch onto current master.

**Comment by Dschee** [comment]
@jpsim I just rebased this, too. Thanks for merging the others!

For an example on how we configure these rules, see [here](https://github.com/JamitLabs/ProjLintTemplates/blob/master/iOS/SwiftLint.stencil#L95) and [there](https://github.com/JamitLabs/ProjLintTemplates/blob/master/iOS/SwiftLint.stencil#L107).

File: Source/SwiftLintFramework/Rules/Style/TypeContentsOrderRule.swift [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+        _ substructure: [String: SourceKitRepresentable],
+        in file: File
+    ) -> [StyleViolation] {
+        let typeContentOffsets = self.typeContentOffsets(in: substructure)
+        let orderedTypeContentOffsets = typeContentOffsets.sorted { lhs, rhs in lhs.offset < rhs.offset }
+
+        var violations =  [StyleViolation]()
+
+        var lastMatchingIndex = -1
+        print(configuration.order)
```
Comment (jpsim) [inline Source/SwiftLintFramework/Rules/Style/TypeContentsOrderRule.swift]: rogue print call

**Comment by Dschee** [comment]
I just rebased this to resolve merge conflicts. @jpsim Would you mind taking another look? I fixed your small comment a week ago, do you have any other reservations or can we merge now (once the CI has passed)?

**Comment by Dschee** [comment]
Just rebased this branch to solve merge conflicts. Any idea when this might be merged @jpsim?

**Comment by Dschee** [comment]
@marcelofabri Would you mind taking a look at this, too when you have time? I know this is a new rule and takes more time to understand. But:

* it's waiting for 6 months now and was rebased 5 times now to keep it up-to-date
* @jpsim already had a cursory look without telling me this is a bad idea
* we are using the rule via my fork in several projects at our company and it works well
* you can see it in action or play with it in the open source projects [BartyCrouch](https://github.com/Flinesoft/BartyCrouch), [HandySwift](https://github.com/Flinesoft/HandySwift) and [HandyUIKit](https://github.com/Flinesoft/HandyUIKit)
* it's opt-in and therefore shouldn't make any troubles with existing projects
* people seem to wait for it since it's the [most upvoted](https://github.com/realm/SwiftLint/pulls?q=is%3Apr+is%3Aopen+sort%3Areactions-%2B1-desc) PR right now

Would be glad to see this merged soon. It opens new ways to enforce Swift style & conventions, which is what this tool is all about.

**Comment by Dschee** [comment]
@marcelofabri @jpsim I don't want to rush it and also don't have a good feeling in merging my own PR, but since you don't seem to have time for further reviewing this and it's opt-in and upvoted multiple times and is tested thouroughly, I'd go for it unless you express any complaints within the next say 10 days.

**Comment by rumnat** [comment]
Thanks for this rule. How to setup? I can't open the configure examples

**Comment by Dschee** [comment]
@rumnat Oh yes, the links changed, I just updated them, they should work again!
(The code is currently commented out since this PR is still not merged, will un-comment once this is merged and a new release was made.)

**Comment by Dschee** [comment]
Okay, just as I announced a month ago (and I waited way longer than 10 days), since there was no feedback whatsoever since then and due to the reasons explained [here](https://github.com/realm/SwiftLint/pull/2296#issuecomment-463150590), I'm merging this without additional feedback from other contributors once CI is passed.

**Comment by rumnat** [comment]
I'd like to have the following structure:
1) enums
2) protocols
3) Main Class
  - outlets
  - private field
  - public fields
  - overriding methods
  - public methods 
4) Private extension of main class
5) Rest of Extensions for each protocol implementation or separate logic

Can I reach it using your rule? Thanks in advance

**Comment by Dschee** [comment]
You can't differentiate between the types as of now (e.g. enums vs. protocols, both would belong to `supporting_type`) also not between `private` and `public`, but try this with the `file_types_order` rule:

```yaml
file_types_order:
  order:
    - supporting_type
    - main_type
    - extension
```

Note that this is actually the default, so you could also just turn on the rule and it should be like this.

Regarding the type contents, what you want is currently not possible as you can't differentiate between `private` and `public` accessor, also you currently can't detect if a method is overriding or not. Those would all be even more detailed and could be added as an improvement to this rule. You can define the kinds of contents though, to me it looks like this might fit your preferences (assuming you first prefer type functions, then object functions):

```yaml
type_contents_order:
  order:
    - case
    - [type_alias, associated_type, subtype]
    - ib_outlet
    - [type_property, ib_inspectable, instance_property]
    - initializer
    - type_method
    - view_life_cycle_method
    - [ib_action, other_method]
    - subscript
```

Feel free to create a issues which request an additional rulse named something like `visibility_modifier_order` and `override_order` where we could enforce a specific modifier order given the type is the same. E.g. specifying something like this:

```
visibility_modifier_order:
  order:
    - [public, open]
    - [internal, implicit]
    - [private, fileprivate]

override_order:
  override_top: true
```

**Comment by piobyz** [comment]
> For an example on how we configure these rules, see [here](https://github.com/JamitLabs/ProjLintTemplates/blob/master/JamitLabs/App/SwiftLint.stencil#L97)

Wow, what a great source of Custom Rules, thanks @Dschee

**Comment by jpsim** [comment]
Thanks for all your work on this PR and really stepping up as a contributor and maintainer of this project overall @Dschee! I especially appreciate how thorough you were in adding tests and considering edges cases here.

> Okay, just as I announced a month ago (and I waited way longer than 10 days), since there was no feedback whatsoever since then and due to the reasons explained [here](https://github.com/realm/SwiftLint/pull/2296#issuecomment-463150590), I'm merging this without additional feedback from other contributors once CI passes.

I understand how frustrating it must be to not get feedback on your work for very long stretches of time. You did the right thing merging this. I'm sorry for not being very involved in SwiftLint on a regular basis any more 😕.

SwiftLint doesn't have a single gatekeeper (I'm sorry if I ever gave that impression). There are over a dozen people with write access to this git repository and **any one of those contributors should feel empowered to merge PRs, cut releases and help review other PRs/issues**.

I hope you'll keep contributing to this project as you've really made it so much better by your contributions so far.

**Comment by Dschee** [comment]
Thank you for those kind words @jpsim and also for approving my self-merge. I still feel bad about it since usually I‘m the guy stressing how important a review is, but I felt different in this case. Thank you for your understanding!

I can understand that maintaining open source software - especially when there are so many users with different needs - can become hard at times. That‘s why I tried to help out when I had some time left over. Let‘s keep improving SwiftLint all together!

**Comment by samrayner** [comment]
Thanks for this awesome addition @Dschee!

I'm getting one false positive. Files that only contain extensions are being flagged with:

> File Types Order Violation: An 'extension' should not be placed amongst the file type(s) 'main_type'. (file_types_order)

**Comment by Dschee** [comment]
@samrayner Could you please open an issue with that false positive so we can track it and fix it?

Also, please provide a code example where this happens so we can properly reproduce it.

**Comment by samrayner** [comment]
@Dschee https://github.com/realm/SwiftLint/issues/2749 Sorry for the delay!

**Comment by Lutzifer** [comment]
great rule. How do we get this to autocorrect?

**Comment by rogerluan** [comment]
Just dropping a line to show my appreciation over this rule! Absolutely great addition, can't wait to use it! 🚀

**Pull request merged**